src/compiler/transformers/classFields.ts

```diff
@@ -506,6 +786,165 @@ namespace ts {
+                                    visitor
+                                )
+                            )]
+                        )
+                    )
+                ]),
+                "value"
+            );
+        }
+
+        function transformDestructuringAssignmentTarget(node: ArrayLiteralExpression | ObjectLiteralExpression) {
+            const hasPrivateIdentifiers = isArrayLiteralExpression(node) ?
+                forEach(node.elements, isPrivateIdentifierPropertyAccessExpression) :
+                forEach(node.properties, property => isPropertyAssignment(property) && isPrivateIdentifierPropertyAccessExpression(property.initializer));
+            if (!hasPrivateIdentifiers) {
+                return node;
+            }
+            if (isArrayLiteralExpression(node)) {
+                // Transforms private names in destructuring assignment array bindings.
+                //
+                // Source:
+                // ([ this.#myProp ] = [ "hello" ]);
+                //
+                // Transformation:
+                // [ { set value(x) { this.#myProp = x; } }.value ] = [ "hello" ];
+                return updateArrayLiteral(
+                    node,
+                    node.elements.map(
+                        expr => isPrivateIdentifierPropertyAccessExpression(expr) ?
+                            wrapPrivateIdentifierForDestructuringTarget(expr) :
+                            expr
+                    )
+                );
+            }
+            else {
+                // Transforms private names in destructuring assignment object bindings.
+                //
+                // Source:
+                // ({ stringProperty: this.#myProp } = { stringProperty: "hello" });
+                //
+                // Transformation:
+                // ({ stringProperty: { set value(x) { this.#myProp = x; } }.value }) = { stringProperty: "hello" };
+                return updateObjectLiteral(
+                    node,
+                    node.properties.map(
+                        prop => isPropertyAssignment(prop) && isPrivateIdentifierPropertyAccessExpression(prop.initializer) ?
+                            updatePropertyAssignment(
+                                prop,
+                                prop.name,
+                                wrapPrivateIdentifierForDestructuringTarget(prop.initializer)
+                            ) :
+                            prop
+                    )
+                );
+            }
+        }
     }
 
+    function createPrivateInstanceFieldInitializer(receiver: LeftHandSideExpression, initializer: Expression | undefined, weakMapName: Identifier) {
+        return createCall(
+            createPropertyAccess(weakMapName, "set"),
+            /*typeArguments*/ undefined,
+            [receiver, initializer || createVoidZero()]
+        );
+    }
+
+    export const classPrivateFieldGetHelper: UnscopedEmitHelper = {

```

Comment:
These should be moved to *factory.ts* and marked `/* @internal */`. In addition, all unscoped helpers should be added to `getAllUnscopedEmitHelpers` in *factory.ts* so that they work well with project reference builds.

Reply:
Why move them? None of the other helpers are declared in factory.ts and it seems specific to this transformation.

Reply:
As long as they are added to `getAllUnscopedEmitHelpers` then they don't need to be moved.